### ISSUE TITLE
surround.vim: dot-repeat

### ISSUE THREAD

**Issue by justinmk** [justinmk/vim-sneak#94]
Repeat doesn't quite work for:

```
ysta)
ysfa)
yszab)
```

**Comment by wilywampa** [comment]
I worked out a solution for this. It requires small modification to both vim-sneak and vim-surround.

**tl;dr** version is make vim-surround set `g:first_op` to 1 the first time its `opfunc` is called and `g:repeat_op` to `&opfunc` the first time it's called and vim-sneak needs to check if it's being used for a repeated `g@` operator and repeat instead of doing a new motion.

This is the change to vim-sneak: https://github.com/wilywampa/vim-sneak/commit/a9a02c2632f7d9906676b791d1de64b136f67c3c
This is the change to vim-surround: https://github.com/wilywampa/vim-surround/commit/9d67d312caa850a1a7bc518de12e83cd2d06a134

Currently, if you do any sneak motion, then repeat it explicitly when doing the surround command, everything works fine. For example, `f)<C-o>ys;aj.` will move the cursor to the next `)`, go back to the start, surround from the start to and including the `)` with `<>`, move the cursor down one line, and surround the new cursor position and the next `)` with `<>`. This works because Vim calls `<Plug>SneakNext` directly.

If, on the other hand, you try to do `ysf)aj.`, the first surround will work, but the second will ask for another character before completing, then surround from the cursor position to and including the next `.` (instead of `)` like it should) with whatever new character you enter. That's because Vim is calling `<Plug>Sneak_t` instead of `<Plug>SneakNext` because it calls whichever was called most recently, which is now `Sneak_t`. Now when vim-sneak goes into `sneak#wrap()`, it starts a new sneak motion instead of repeating the old one. sneak then eats the `)` character from `g:repeat_sequence` (which is currently `<Plug>SurroundRepeat)` (note the extra parenthesis at the end, which is supposed to be used by vim-surround but vim-repeat gets it instead)) in `s:getnchars()`.

The solution I came up with is to add logic to `sneak#wrap()` to check if it's being called as part of a repeated `opfunc`. The first time `sneak#wrap()` is called (at `ysf`) needs to go through the `sneak#to()` machinery like normal, but subsequent repeats should go straight to `sneak#repeat()`. This is where the change to vim-surround comes in. As far as I can tell, there is no way to tell if an `opfunc` is being called for the first time or being repeated, so vim-surround needs to put out a flag saying that it needs a new motion instead of a repeat motion.

The two flags (`g:first_op` and `g:repeat_op`) might seem redundant, but they're necessary in case vim-sneak is being used with other `opfunc`s. vim-sneak needs to know if vim-surround (or whatever other plugin wants to use the two flags) is still the current `opfunc` or else behave like it should behave like a new `opfunc` with a new motion. Note that the way `g:first_op` is set makes two different sneak motions consecutively still work as expected (e.g. `ysf)bjyst.]` or whatever).

**Comment by justinmk** [comment]
@wilywampa Thanks for sharing that; now at least there is a known boundary to the problem. I don't think @tpope (nor I) will like adding a global flag for this purpose, but it gives me something to work with.

**Comment by wilywampa** [comment]
I found a fix without any globals or modifications outside of vim-sneak. The catch is it requires Vim 7.4.786 or newer. Patch 786 introduced the `OptionSet` autocommand: http://vimhelp.appspot.com/autocmd.txt.html#OptionSet

Here's another branch I made that uses the new autocommand and works just as well as my other solution: https://github.com/wilywampa/vim-sneak/commit/e675e714bcf291f8aada7c0831edb9891f75d2fa

For some reason `exists('#OptionSet')` doesn't work so I had to do version checking the ugly way.

**Comment by justinmk** [comment]
Nice! Thanks again.

**Comment by jwmann** [comment]
Has this change been merged in? I constantly find myself using the `ys` method for vim-surround and then realizing I need to select it with a V-block first because Sneak takes precedence.

**Comment by justinmk** [comment]
AFAIU it would require a change to vim-surround. So there's nothing I can do about it.

**Comment by brendanator** [comment]
@justinmk @jwmann I suggest reopening and merging PR #219 

I've just checked it out and tested it and it resolves this issue without any changes to vim-surround

**Comment by jwmann** [comment]
Are you sure? @justinmk said it needed a change. :S

**Comment by brendanator** [comment]
See @wilywampa comment on 30 Jul 2015

     I found a fix without any globals or modifications outside of vim-sneak

Try it out and see

**Comment by jwmann** [comment]
I haven't tried it out just yet because I don't have a lot of time but this commit seems to suggest it needs a change in Surround

wilywampa/vim-surround@9d67d31

If only @wilywampa himself could clear this up for us 😅

**Comment by justinmk** [comment]
@jwmann i reopened it myself in https://github.com/justinmk/vim-sneak/pull/224 but I haven't had time to write a test, nor stare at the code. If someone wants to write a test that will surely help.

**Comment by jwmann** [comment]
Thanks! I'll test it myself on the weekend! :smile_cat:
